### PR TITLE
Remove the requirement that ast->hir lowering be reproducible

### PR DESCRIPTION
**Pull request by jseyfried**
This PR changes the ast->hir lowerer to be non-reproducible, and it removes the lowering context's id cache.

If the `hir` of an `ast` node needs to be reproduced, we can use the hir map instead of the lowerer -- for example, `tcx.map.expect_expr(expr.id)` instead of `lower_expr(lcx, expr)`.

r? @nrc

### CODE REVIEW

**Comment by nagisa** [comment]
What’s the purpose of the change? Seems like a potentially plugin-breaking-change to me.

**Comment by petrochenkov** [comment]
Side note: if `librustc_save_analysis` doesn't use AST -> HIR lowering anymore, then MTWT tables can be cleared even if save_analysis is enabled. See https://github.com/rust-lang/rust/pull/30145#discussion_r46775767

**Comment by jseyfried** [comment]
@nagisa 
> What’s the purpose of the change?

The purpose is to simplify lowering (by removing the id caching system) and to remove the lowering context from the [`CompileState`](https://github.com/rust-lang/rust/blob/b0aefff71499ac8e5f65c4a241b279e687c44c36/src/librustc_driver/driver.rs#L334).

Also, it will be significantly simpler to do name resolution during lowering if lowering doesn't need to be idempotent. I don't think it's worth the extra complexity needed to be idempotent when we can just use the HIR map to re-lower (almost) any AST node.

**Comment by jseyfried** [comment]
@petrochenkov good point, I added a commit that always clears the MTWT tables after lowering.

File: src/librustc/hir/lowering.rs [hunk 1]
```diff
@@ -2121,21 +2121,8 @@ fn signal_block_expr(lctx: &LoweringContext,
 
 #[cfg(test)]
 mod test {
```
Comment (nrc) [inline src/librustc/hir/lowering.rs]: I think you can remove this whole module

File: src/librustc/hir/lowering.rs [hunk 2]
```diff
@@ -83,13 +68,8 @@ use std::cell::{Cell, RefCell};
...
     crate_root: Option<&'static str>,
-    // Map AST ids to ids used for expanded nodes.
-    id_cache: RefCell<HashMap<NodeId, NodeId>>,
     // Use if there are no cached ids for the current node.
     id_assigner: &'a NodeIdAssigner,
-    // 0 == no cached id. Must be incremented to align with previous id
-    // incrementing.
-    cached_id: Cell<u32>,
     // Keep track of gensym'ed idents.
     gensym_cache: RefCell<HashMap<(NodeId, &'static str), hir::Ident>>,
```
Comment (nrc) [inline src/librustc/hir/lowering.rs]: You should be able to remove the gensym stuff too, iirc, that is only used for reproducible lowering.

File: src/librustc/hir/lowering.rs [hunk 3]
```diff
@@ -83,13 +68,8 @@ use std::cell::{Cell, RefCell};
 
 pub struct LoweringContext<'a> {
     crate_root: Option<&'static str>,
-    // Map AST ids to ids used for expanded nodes.
-    id_cache: RefCell<HashMap<NodeId, NodeId>>,
     // Use if there are no cached ids for the current node.
```
Comment (nrc) [inline src/librustc/hir/lowering.rs]: COmment needs updating

File: src/librustc/hir/lowering.rs [hunk 4]
```diff
@@ -201,47 +171,12 @@ impl<'a, 'hir> LoweringContext<'a> {
 fn cache_ids<'a, OP, R>(lctx: &LoweringContext, expr_id: NodeId, op: OP) -> R
```
Comment (nrc) [inline src/librustc/hir/lowering.rs]: should be able to remove this function entirely

File: src/librustc/session/config.rs [hunk 5]
```diff
@@ -664,8 +664,6 @@ options! {DebuggingOptions, DebuggingSetter, basic_debugging_options,
           "for every macro invocation, print its name and arguments"),
     enable_nonzeroing_move_hints: bool = (false, parse_bool,
           "force nonzeroing move optimization on"),
-    keep_mtwt_tables: bool = (false, parse_bool,
```
Comment (nrc) [inline src/librustc/session/config.rs]: you should keep this option

File: src/librustc_driver/driver.rs [hunk 6]
```diff
@@ -245,12 +243,6 @@ pub fn compile_input(sess: &Session,
     Ok(())
 }
 
-fn keep_mtwt_tables(sess: &Session) -> bool {
-    sess.opts.debugging_opts.keep_mtwt_tables ||
-    sess.opts.debugging_opts.save_analysis ||
-    sess.opts.debugging_opts.save_analysis_csv
```
Comment (nrc) [inline src/librustc_driver/driver.rs]: just remove the save-analysis checks from here

File: src/librustc/hir/lowering.rs (hunk 2 above)
Comment (jseyfried) [inline src/librustc/hir/lowering.rs]: Will do

File: src/librustc/session/config.rs (hunk 5 above)
Comment (jseyfried) [inline src/librustc/session/config.rs]: Will do

File: src/librustc/hir/lowering.rs (hunk 3 above)
Comment (jseyfried) [inline src/librustc/hir/lowering.rs]: Will do

File: src/librustc/hir/lowering.rs (hunk 1 above)
Comment (jseyfried) [inline src/librustc/hir/lowering.rs]: Done

File: src/librustc/hir/lowering.rs (hunk 4 above)
Comment (jseyfried) [inline src/librustc/hir/lowering.rs]: Done

File: src/librustc_driver/driver.rs (hunk 6 above)
Comment (jseyfried) [inline src/librustc_driver/driver.rs]: Done

File: src/librustc/hir/lowering.rs [hunk 7]
```diff
@@ -29,21 +29,6 @@
...
-// using node ids starting with the cached id, we ensure ids are reproducible.
-// To use this system, you just need to hold on to a CachedIdSetter object
-// whilst lowering. This is an RAII object that takes care of setting and
-// restoring the cached id, etc.
-//
-// This whole system relies on node ids being incremented one at a time and
-// all increments being for lowering. This means that you should not call any
-// non-lowering function which will use new node ids.
-//
 // We must also cache gensym'ed Idents to ensure that we get the same Ident
```
Comment (nrc) [inline src/librustc/hir/lowering.rs]: This comment can probably go too

**Comment by nrc** [comment]
r+ with that comment removed

File: src/librustc/hir/lowering.rs (hunk 7 above)
Comment (jseyfried) [inline src/librustc/hir/lowering.rs]: Good point, forgot about that.

Comment (jseyfried) [inline src/librustc/hir/lowering.rs]: Done.

**Comment by jseyfried** [comment]
@nrc Thanks!
I'm almost done with a PR for resolving AST, pending #33190.

**Comment by RalfJung** [comment]
What exactly is getting non-reproducible with this change? Does this mean that if I compile the same Rust program twice, I would get two different binaries (as in, binaries that are not bit-for-bit identical)? That would be a huge bummer -- reproducible builds are a cornerstone of building trustworthy binaries. If Rust can't build reproducibly, This is actively used, for example, by the Tor project and some BitCoin clients. it would be unsuited for any application that needs to establish trust in the generated binaries. See <https://reproducible-builds.org/> for more details.

**Comment by jseyfried** [comment]
@RalfJung 
> Does this mean that if I compile the same Rust program twice, I would get two different binaries (as in, binaries that are not bit-for-bit identical)?

I'm not sure if Rust has this property, but this PR isn't related to reproducible builds. Lowering is still deterministic after this PR (i.e. it behaves the same when building the same code).

This PR changed lowering so that if the same AST node is lowered multiple times *during the same build*, we aren't guaranteed to get same result. Perhaps a better word would be "idempotent" instead of "reproducible".

**Pull request merged**